command/agent/command.go

```diff
@@ -295,13 +295,26 @@ func (c *Command) setupAgent(config *Config, logOutput io.Writer, logWriter *log
 		return err
 	}
 
-	rpcListener, err := net.Listen("tcp", rpcAddr.String())
+	if _, ok := rpcAddr.(*net.UnixAddr); ok {
+		// Remove the socket if it exists, or we'll get a bind error
+		_ = os.Remove(rpcAddr.String())
```

Comment:
If we deferred removal of the socket after the Listen, it'd kill the socket when setupAgent exited. But we don't need to do that, because the file is removed automatically by Go when the socket is closed. The code here is to catch the times when that doesn't happen -- for instance if consul dies unexpectedly. We need to ensure it's gone before we try listening again, or we'll error out.

You seem to be arguing that erroring out is the right thing to do, and I can see that point, but I don't (necessarily) agree. The user providing a desired path for the socket file indicates that they expect it to be used by consul; if something else is using it, that's the user's fault. I think consul should be able to manage the state of that socket, because the user told it to...including taking it over if it appears to be used by something else (or something that is now dead). Otherwise you force anyone using Unix sockets with consul to write helper scripts to ensure that the socket is cleaned up anytime consul dies, lest consul is unable to start again when (runit, supervisord, upstart, systemd) tries re-launching it.